### PR TITLE
Partially infer conv return types

### LINKED ISSUE #47057: Conv2D output shape becomes fully dynamic when only input batch size is dynamic in tf-nightly
### 1. System information

- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Colab
- TensorFlow installation (pip package or built from source): pip
- TensorFlow library (version, if pip package or github SHA, if built from source): 2.4.1 v.s. nightly

### 2. Code

nightly: https://colab.research.google.com/drive/1r-vzh6JCIc5A4K8aYEbMZANWVrtSkKrF?usp=sharing
2.4.1: https://colab.research.google.com/drive/1ODMv9joqc2ybsSdNHhGq1RWKXYsnFjd-?usp=sharing

### 3. Failure after conversion

Models are correct, but nightly has fully dynamic output shape and 2.4.1 has dynamic batch dim only.

Nightly
image

2.4.1
image

1 here means dynamic dim (netron cannot show it precisely). It does not really matter in my use case though. Just a regression issue.

### PR DESCRIPTION
**Pull request by WindQAQ**
Fixes #47057.

af9ad9d introduces conv inferReturnTypes, but skips the inference when either input or filter does not have static shape (unranked or not all dimensions are static). However, it introduces the regression when it comes to partially dynamic input. This PR instead tries to infer as many dimensions as possible, which avoids shape information loss when there are only some dynamic dimensions (like batch dim).

### CODE REVIEW

File: tensorflow/compiler/mlir/tensorflow/tests/tf-ops.mlir [hunk 1]
```diff
@@ -558,10 +558,90 @@ func @testValidConv2D(%arg0: tensor<256x32x32x3xf32>, %arg1: tensor<3x3x3x16xf32
 
 // -----
 
-// CHECK-LABEL: func @testValidDynamicConv2D
-func @testValidDynamicConv2D(%arg0: tensor<*xf32>, %arg1: tensor<*xf32>) -> tensor<?x?x?x?xf32> {
```
Comment (abattery) [inline tensorflow/compiler/mlir/tensorflow/tests/tf-ops.mlir:562]: Could you add a test case for only dynamic dimension on the batch dim to verify the inference result?

**Review by abattery — APPROVED** [review]
Thanks for your contribution!

File: tensorflow/compiler/mlir/tensorflow/tests/tf-ops.mlir (hunk 1 above)
Comment (WindQAQ) [reply tensorflow/compiler/mlir/tensorflow/tests/tf-ops.mlir:562]: Thanks for the review. See `testDynamicBatchConv2D`.

Comment (abattery) [reply tensorflow/compiler/mlir/tensorflow/tests/tf-ops.mlir:562]: How about moving these new tests under shape inference test if it is available?

Comment (abattery) [reply tensorflow/compiler/mlir/tensorflow/tests/tf-ops.mlir:562]: For example, https://github.com/tensorflow/tensorflow/blob/master/tensorflow/compiler/mlir/tensorflow/tests/shape_inference.mlir

**Comment by miaout17** [comment]
@smit-hinsu Do you want to take a look since you approved #44022 ?

File: tensorflow/compiler/mlir/tensorflow/tests/tf-ops.mlir (hunk 1 above)
Comment (WindQAQ) [reply tensorflow/compiler/mlir/tensorflow/tests/tf-ops.mlir:562]: Thanks for pointing out :) Let me try it real quick.

Comment (WindQAQ) [reply tensorflow/compiler/mlir/tensorflow/tests/tf-ops.mlir]: Updated 😃

File: tensorflow/compiler/mlir/tensorflow/tests/shape_inference.mlir [hunk 2]
```diff
@@ -1045,4 +1045,94 @@ module attributes {tf.versions = {bad_consumers = [], min_consumer = 0 : i32, pr
     // CHECK-SAME: tensor<i32>
     return %arg0 : tensor<*xi32>
   }
+
+  // Test conv2d inferReturnTypes.
```
Comment (abattery) [inline tensorflow/compiler/mlir/tensorflow/tests/shape_inference.mlir:1049]: Nit: Could you modify this sentence to make a complete sentence?

**Review by abattery — APPROVED** [review]

File: tensorflow/compiler/mlir/tensorflow/tests/shape_inference.mlir (hunk 2 above)
Comment (WindQAQ) [reply tensorflow/compiler/mlir/tensorflow/tests/shape_inference.mlir]: See if this is better 😃

**Review by abattery — APPROVED** [review]

**Pull request merged**